starlite/contrib/repository/testing/generic_mock_repository.py

```diff
@@ -0,0 +1,346 @@
+        """Update instances with the attribute values present on ``data``.
+
+        Args:
+            data: A list of instances that should have a value for :attr:`id_attribute <GenericMockRepository.id_attribute>` that exists in the
+                collection.
+
+        Returns:
+            The updated instances.
+
+        Raises:
+            RepositoryNotFoundException: If no instance found with same identifier as ``data``.
+        """
+        items = [self._find_or_raise_not_found(self.get_id_attribute_value(row)) for row in data]
+        # should never be modifiable
+        for item in items:
+            if hasattr(item, "updated"):
+                # maybe the @declarative_mixin decorator doesn't play nice with pyright?
+                item.updated = datetime.now()  # pyright: ignore # noqa: DTZ005
+            for key, val in item.__dict__.items():
+                if key.startswith("_"):
+                    continue
+                setattr(item, key, val)
+        return items
+
+    async def upsert(self, data: ModelT) -> ModelT:
+        """Update or create instance.
+
+        Updates instance with the attribute values present on ``data``, or creates a new instance if
+        one doesn't exist.
+
+        Args:
+            data: Instance to update existing, or be created. Identifier used to determine if an
+                existing instance exists is the value of an attribute on `data` named as value of
+                :attr:`id_attribute <GenericMockRepository.id_attribute>`.
+
+        Returns:
+            The updated or created instance.
+
+        Raises:
+            RepositoryNotFoundException: If no instance found with same identifier as ``data``.
+        """
+        item_id = self.get_id_attribute_value(data)
+        if item_id in self.collection:
+            return await self.update(data)
+        return await self.add(data, allow_id=True)
+
+    async def list_and_count(
+        self,
+        *filters: FilterTypes,
+        **kwargs: Any,
+    ) -> tuple[list[ModelT], int]:
+        """Get a list of instances, optionally filtered with a total row count.
+
+        Args:
+            *filters: Types for specific filtering operations.
+            **kwargs: Instance attribute value filters.
+
+        Returns:
+            Count of records returned by query, ignoring pagination.
+        """
+        return await self.list(*filters, **kwargs), await self.count(*filters, **kwargs)
+
+    async def list(self, *filters: "FilterTypes", **kwargs: Any) -> list[ModelT]:
+        """Get a list of instances, optionally filtered.
+
+        Args:
+            *filters: Types for specific filtering operations.
+            **kwargs: Instance attribute value filters.
+
+        Returns:
+            The list of instances, after filtering applied.
+        """
+        return list(self.filter_collection_by_kwargs(self.collection, **kwargs).values())
+
+    def filter_collection_by_kwargs(  # type:ignore[override]
```

Comment:
wonder why the override ignore is necessary here, the signature looks right to me compared to the abc... :shrug:

Reply:
this is weird to me, the super class has an unconstrained type var as the collection arg type and returns same. IMO, this is exactly what we are implementing here.